Add support for Python 3.12

It would be great to add Python 3.12 support in the near future. Currently installs fail for me with: 


```bash  • Installing pyg4ometry (1.1.2): Failed

  ChefBuildError

  Backend subprocess exited when trying to invoke build_wheel
  
  /tmp/tmptbzg1t85/.venv/lib/python3.12/site-packages/setuptools_scm/git.py:308: UserWarning: git archive did not support describe output
    warnings.warn("git archive did not support describe output")
  /tmp/tmptbzg1t85/.venv/lib/python3.12/site-packages/setuptools_scm/git.py:327: UserWarning: unprocessed git archival found (no export subst applied)
    warnings.warn("unprocessed git archival found (no export subst applied)")
  *** scikit-build-core 0.7.1 using CMake 3.28.1 (wheel)
  *** Configuring CMake...
  loading initial cache file /tmp/tmp4t3tkzsh/build/CMakeInit.txt
  -- The C compiler identification is GNU 10.2.1
  -- The CXX compiler identification is GNU 10.2.1
  -- Detecting C compiler ABI info
  -- Detecting C compiler ABI info - done
  -- Check for working C compiler: /usr/bin/cc - skipped
  -- Detecting C compile features
  -- Detecting C compile features - done
  -- Detecting CXX compiler ABI info
  -- Detecting CXX compiler ABI info - done
  -- Check for working CXX compiler: /usr/bin/c++ - skipped
  -- Detecting CXX compile features
  -- Detecting CXX compile features - done
  -- Setting build type to 'Release'
  -- Found Python: /tmp/tmptbzg1t85/.venv/bin/python (found version "3.12.0") found components: Interpreter Development.Module 
  -- Performing Test HAS_FLTO
  -- Performing Test HAS_FLTO - Success
  -- Found pybind11: /tmp/tmptbzg1t85/.venv/lib/python3.12/site-packages/pybind11/include (found version "2.11.1")
  CMake Warning (dev) at cmake/modules/FindPythonExtensions.cmake:247 (find_package):
    Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
    are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
    the cmake_policy command to set the policy and suppress this warning.
  
  Call Stack (most recent call first):
    CMakeLists.txt:33 (find_package)
  This warning is for project developers.  Use -Wno-dev to suppress it.
  
  -- Found PythonInterp: /tmp/tmptbzg1t85/.venv/bin/python (found version "3.12") 
  CMake Warning (dev) at cmake/modules/FindPythonExtensions.cmake:248 (find_package):
    Policy CMP0148 is not set: The FindPythonInterp and FindPythonLibs modules
    are removed.  Run "cmake --help-policy CMP0148" for policy details.  Use
    the cmake_policy command to set the policy and suppress this warning.
  
  Call Stack (most recent call first):
    CMakeLists.txt:33 (find_package)
  This warning is for project developers.  Use -Wno-dev to suppress it.
  
  -- Found PythonLibs: /usr/local/lib/libpython3.12.so (found version "3.12.0") 
  -- Found Cython: /tmp/tmptbzg1t85/.venv/bin/cython  
  CMake Error at CMakeLists.txt:37 (find_package):
    By not providing "FindCGAL.cmake" in CMAKE_MODULE_PATH this project has
    asked CMake to find a package configuration file provided by "CGAL", but
    CMake did not find one.
  
    Could not find a package configuration file provided by "CGAL" (requested
    version 5.0) with any of the following names:
  
      CGALConfig.cmake
      cgal-config.cmake
  
    Add the installation prefix of "CGAL" to CMAKE_PREFIX_PATH or set
    "CGAL_DIR" to a directory containing one of the above files.  If "CGAL"
    provides a separate development package or SDK, be sure it has been
    installed.
  
  
  -- Configuring incomplete, errors occurred!
  
  *** CMake configuration failed
  

  at /usr/local/py-utils/venvs/poetry/lib/python3.12/site-packages/poetry/installation/chef.py:164 in _prepare
      160│ 
      161│                 error = ChefBuildError("\n\n".join(message_parts))
      162│ 
      163│             if error is not None:
    → 164│                 raise error from None
      165│ 
      166│             return path
      167│ 
      168│     def _prepare_sdist(self, archive: Path, destination: Path | None = None) -> Path:

Note: This error originates from the build backend, and is likely not a problem with poetry but with pyg4ometry (1.1.2) not supporting PEP 517 builds. You can verify this by running 'pip wheel --no-cache-dir --use-pep517 "pyg4ometry (==1.1.2)"'.```

--- a/.github/workflows/distribution.yml
+++ b/.github/workflows/distribution.yml
@@ -44,7 +44,7 @@ jobs:
           MAKEFLAGS: -j${{ steps.cpu-cores.outputs.count }}
 
       - name: Upload wheels
-        uses: actions/upload-artifact@v4
+        uses: actions/upload-artifact@v3
         with:
           path: wheelhouse/*.whl
 
@@ -53,12 +53,12 @@ jobs:
     runs-on: ubuntu-latest
     if: github.event_name == 'release' && github.event.action == 'published'
     steps:
-      - uses: actions/download-artifact@v4
+      - uses: actions/download-artifact@v3
         with:
           name: artifact
           path: dist
 
-      - uses: pypa/gh-action-pypi-publish@v1.8.11
+      - uses: pypa/gh-action-pypi-publish@release/v1
         with:
           user: __token__
           password: ${{ secrets.pypi_password }}

--- a/.github/workflows/main.yml
+++ b/.github/workflows/main.yml
@@ -19,7 +19,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.8", "3.9", "3.10", "3.11"]
+        python-version: ["3.8", "3.9", "3.10", "3.11", "3.12"]
         os: [ubuntu-latest, macOS-latest]
     steps:
       - uses: actions/checkout@v4

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -130,7 +130,6 @@ extend-ignore = [
   "PT004",  # Use underscore for non-returning fixture (use usefixture instead)
   "RUF012", # Mutable class attributes should be annotated with typing.ClassVar
 ]
-target-version = "py37"
 src = ["src"]
 unfixable = [
   "T20",  # Removes print statements

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -91,7 +91,7 @@ metadata.version.provider = "scikit_build_core.metadata.setuptools_scm"
 minversion = "6.0"
 addopts = ["-ra", "--showlocals", "--strict-markers", "--strict-config"]
 xfail_strict = true
-filterwarnings = "error"
+filterwarnings = ["error", "ignore::DeprecationWarning"]
 log_cli_level = "info"
 testpaths = "tests"
 

--- a/pyproject.toml
+++ b/pyproject.toml
@@ -151,5 +151,5 @@ manylinux-x86_64-image = "g4edge/manylinux2014_x86_64:latest"
 archs = ["auto"]
 
 [tool.cibuildwheel.macos]
-archs = ["x86_64", "arm64"]
+archs = ["x86_64"]
 before-all = ".github/bin/install-deps-macos.sh"

